lib/Jsrt/Jsrt.cpp

```diff
@@ -4266,3 +4161,431 @@ CHAKRA_API JsTTDReplayExecution(_Inout_ JsTTDMoveMode* moveMode, _Inout_ int64_t
+    _Out_ JsValueRef *result)
+{
+    return CompileRun(scriptVal, sourceContext, sourceUrl, parseAttributes,
+        result, false);
+}
+
+CHAKRA_API JsCreatePropertyIdUtf8(
+    _In_z_ const char *name,
+    size_t length,
+    _Out_ JsPropertyIdRef *propertyId)
+{
+    PARAM_NOT_NULL(name);
+    utf8::NarrowToWide wname(name, length);
+    if (!wname)
+    {
+        return JsErrorOutOfMemory;
+    }
+
+    return JsGetPropertyIdFromName(wname, propertyId);
+}
+
+CHAKRA_API JsWritePropertyIdUtf8(
+    JsPropertyIdRef propertyId,
+    uint8_t* buffer,
+    size_t bufferSize,
+    size_t* length)
+{
+    PARAM_NOT_NULL(propertyId);
+
+    const char16* str = nullptr;
+    JsErrorCode errorCode = JsGetPropertyNameFromId(propertyId, &str);
+
+    if (errorCode != JsNoError)
+    {
+        return errorCode;
+    }
+
+    utf8::WideToNarrow utf8Str(str);
+    if (!buffer)
+    {
+        if (length)
+        {
+            *length = utf8Str.Length();
+        }
+    }
+    else
+    {
+        size_t count = min(bufferSize, utf8Str.Length());
+        // Try to copy whole characters if buffer size insufficient
+        auto maxFitChars = utf8::ByteIndexIntoCharacterIndex(
+            (LPCUTF8)(const char*)utf8Str, count,
+            utf8::DecodeOptions::doChunkedEncoding);
+        count = utf8::CharacterIndexToByteIndex(
+            (LPCUTF8)(const char*)utf8Str, utf8Str.Length(), maxFitChars);
+
+        memmove(buffer, utf8Str, sizeof(uint8_t) * count);
+        if (length)
+        {
+            *length = count;
+        }
+    }
+
+    return JsNoError;
+}
+
+CHAKRA_API JsSerialize(
+    JsValueRef scriptVal,
+    BYTE *buffer,
+    unsigned int *bufferSize,
+    JsParseScriptAttributes parseAttributes)
+{
+    PARAM_NOT_NULL(scriptVal);
+    VALIDATE_JSREF(scriptVal);
+
+    bool isExternalArray = Js::ExternalArrayBuffer::Is(scriptVal),
+         isString = false;
+    bool isUtf8   = !(parseAttributes & JsParseScriptAttributeUtf16Encoding);
+    if (!isExternalArray)
+    {
+        isString = Js::JavascriptString::Is(scriptVal);
+        if (!isString)
+        {
+            return JsErrorInvalidArgument;
+        }
+    }
+
+    LoadScriptFlag scriptFlag;
+    const byte* script = isExternalArray ?
+        ((Js::ExternalArrayBuffer*)(scriptVal))->GetBuffer() :
+        (const byte*)((Js::JavascriptString*)(scriptVal))->GetSz();
+    const size_t cb = isExternalArray ?
+        ((Js::ExternalArrayBuffer*)(scriptVal))->GetByteLength() :
+        ((Js::JavascriptString*)(scriptVal))->GetLength();
+
+    if (isExternalArray && isUtf8)
+    {
+        scriptFlag = (LoadScriptFlag) (LoadScriptFlag_ExternalArrayBuffer | LoadScriptFlag_Utf8Source);
+    }
+    else if (isUtf8)
+    {
+        scriptFlag = (LoadScriptFlag) (LoadScriptFlag_Utf8Source);
+    }
+    else
+    {
+        scriptFlag = LoadScriptFlag_None;
+    }
+
+    return JsSerializeScriptCore(script, cb, scriptFlag, nullptr,
+        0, buffer, bufferSize, scriptVal);
```

Comment:
Some of JsParseScriptAttributes are not transferred to LoadScriptFlag like LbraryCode.